crates/cxx-qt-gen/src/naming/mod.rs

```diff
@@ -29,24 +29,76 @@ use crate::syntax::{
 
+/// This struct contains all names a certain syntax element may have
+///
+/// This includes the rust_name, cxx_name, as well as qualifications like
+/// the rust module and C++ namespace.
+#[derive(Debug)]
+pub struct Name {
+    rust: Ident,
+    cxx: Option<String>,
+    module: Ident,
+    namespace: Option<String>,
+}
+
+impl Name {
+    fn from_ident_and_attrs(
+        ident: &Ident,
+        attrs: &[Attribute],
+        parent_namespace: Option<&str>,
+        module: &Ident,
+    ) -> Result<Self> {
+        // Find if there is a namespace (for C++ generation)
+        let mut namespace = if let Some(index) = attribute_find_path(attrs, &["namespace"]) {
+            Some(expr_to_string(
+                &attrs[index].meta.require_name_value()?.value,
+            )?)
+        } else {
+            parent_namespace.map(|namespace| namespace.to_owned())
+        };
+
+        // This is an important check as it allows for the namespace to be cleared by assigning an
+        // empty namespace (i.e. #[namespace = ""])
+        if let Some(namespace_name) = &namespace {
+            if namespace_name.is_empty() {
+                namespace = None;
+            }
+        }
+
+        // Find if there is a cxx_name mapping (for C++ generation)
+        let cxx_name = attribute_find_path(attrs, &["cxx_name"])
+            .map(|index| -> Result<_> {
+                expr_to_string(&attrs[index].meta.require_name_value()?.value)
+            })
+            .transpose()?;
+
+        // Find if there is a rust_name mapping
+        let rust_ident = if let Some(index) = attribute_find_path(attrs, &["rust_name"]) {
+            format_ident!(
+                "{}",
+                expr_to_string(&attrs[index].meta.require_name_value()?.value)?,
+                span = attrs[index].span()
+            )
+        } else {
+            ident.clone()
+        };
+
+        Ok(Self {
+            rust: rust_ident,
+            cxx: cxx_name,
+            namespace,
+            module: module.clone(),
+        })
+    }
+}
+
 /// The purpose of this struct is to store all nameable types.
 ///
 /// This is used by the generator phase to find types by their identifier, such that they can be
 /// fully qualified in Rust and C++.
 #[derive(Default, Debug)]
 pub struct TypeNames {
-    /// Map of the cxx_name of any types defined in CXX extern blocks
-    ///
-    /// This is used in the C++ generation to map the Rust type name to the C++ name
-    cxx_names: BTreeMap<String, String>,
-    /// Map of the namespace of any types or methods defined in CXX extern blocks
-    ///
-    /// This is used in the C++ generation to map the Rust type name to the C++ name
-    namespaces: BTreeMap<String, String>,
-    /// Mappings for CXX types when used outside the bridge
-    ///
-    /// This is used in the Rust generation to map the bridge type A to ffi::B
-    qualified: BTreeMap<Ident, Path>,
+    names: BTreeMap<Ident, Name>,
```

Comment:
We might want to do some sharing between names in the future (e.g. introduce a single `bridge_namespace`) or something like that for efficiency reasons.

Also, I think we can't add new methods to a type alias, right?

Reply:
I thought you could ... If you think there are going to be more fields then it's fine. There is also the new type route otherwise `struct TypeNames(BTreeMap<Ident, Name>)`

Reply:
Right.
These are all pretty much equivalent options I'd say.

Reply:
For now let's stick with what we have, it's encapsulated so doesn't matter which option we choose